### ISSUE TITLE
Fix Unit of Work

### ISSUE THREAD

**Issue by einari** [Cratis/Chronicle#1405]
The purpose of the design of the unit of work has been to be agnostic to what is working with it.
This is to allow you to basically create a unit of work and anything that produces events will then just register events of there is an active unit of work.

Today we only have implementation for AggregateRoot, but having the ability for this for appending events directly on the `IEventLog` would also be very nice.
There are scenarios where you don’t want or don’t need an AggregateRoot and would rather just append directly but in a transactional manner.

We don’t want to support this scenario right now, we just design for it and implement the building blocks we need to cover AggregateRoot scenario right now.

In addition to this we have 2 worlds in which Unit of Work should work; base .NET client - non clustered and the Orleans client, which is Grain based and can be clustered.

These are vastly different in nature. In an in-process, standard TaskScheduler world, we can simply get to the current Unit of Work through an AsyncLocal, this is not possible in the Orleans world where the TaskScheduler gives a new AsyncLocal context and also the AggregateRoot is not necessarily in the same process so it wouldn’t matter if you could get to the same AsyncLocal context.

From the [ASP.NET](http://asp.net/)  perspective a Unit of Work is associated with a web request. Every operation that happens within the scope of this web request will produce results into the Unit of Work, at the end of the request the Unit of Work will automatically commit.

For Orleans, the AggregateRoot is a Grain which means it needs special management.

We have the the following goals:

1. The Unit of Work needs to be owned by the process and AsyncLocal context that starts it
2. Transactional objects needs to be enlisted in the Unit of Work
3. Transactional objects produce / append events that will be captured in the Unit of Work
4. Enlisted objects needs to be told when the Unit of Work is completed - with necessary information

## Incoming call filter
This happens where the Grain being called lives on the incoming of the call.
It should establish the local unit of work through the unit of work manager. 
The result in the local unit of work should be put on the RequestContext using `RequestContext.Set()` after doing `,Invoke()`
on the call context.

## Outgoing call filter
The outgoing call filter is used on the way out for a call. 
After performing the `.Invoke()` on the call context, the resulting events that was put
on the `RequestContext` in the incoming call filter should then be available
through `RequestContext.Get()`. 

These would then be added back to the current unit of work by getting it by its CorrelationId from the
unit of work manager. We probably would need to check if it is the same instance as what we have
in our current AsynLocal context, we don’t need to do anything. 
But if it is different, we should then take the events from the `RequestContext` and put them back
to the root unit of work.

The unit of work will then be committed in the original context.

## Enlisting and flow
We should have an interface for representing types that can be enlisted, something like `IUnitOfWorkEnlistee`.
The UnitOfWork should then have a `Enlist` method that can take a `IUnitOfWorkEnlistee` and add it to a list of `enlistments`. It will keep **enlistments** and at the completion of the UnitOfWork it could call `.Done()` on all the enlistees in the **enlistments** and pass along itself (UnitOfWork) or enough context for the enlistee to perform operations based on the unit of work. Typically for the Orleans AggregateRoot it needs to know the last sequence number of the appended event. So an option here is to introduce a `UnitOfWorkResult` object that keeps the information needed and that we can expand on later.

Typically, the AggregateRoot would be an enlistee and implement the `IUnitOfWorkEnlistee` interface.

Below is an approximation of a diagram showing potentially how this could flow.

![image](https://github.com/user-attachments/assets/3cf0df99-2f24-45a4-b389-e9b958ec094c)

The drawing talks about a `AggregateRootEnlistee`. In non-Orleans we can just let the `AggregateRoot` be an enlistee itself. In-fact, even in Orleans the `AggregateRoot` can be an `Enlistee`. We just probably need a special one for the flow to work in Orleans that handles the coordination. If it can be done without a specific `AggregateRootEnlistee`, that would be simpler and better and preferred. I think it should be possible, for instance the base `IAggregateRoot` interface used for both .NET regular client and Orleans could have the `IUnitOfWorkEnlistee` interface on them and this should probably then work.

**Comment by einari** [comment]
This would also fix #1311

**Comment by woksin** [comment]
How about the scenarios that are outside the scope of an http context, both for orleans and non-orleans aggregates.

And also in terms of internal state of stateless aggregate roots accross different silos, how can the state be consolidated when the unit of work can only be owned by a single process? All the enlistees and their states needs to be on one and the same process / unit of work, right?

**Comment by einari** [comment]
The thinking with this is that everything is based on the `UnitOfWorkManager` that means if we are outside of an HTTP context, everything still applies. You just wouldn't have the automatic begin. This is how it works today. 

For state consolidation. There are potentially multiple `UnitOfWork` instances at play here. One on each silo (process), but there is only one that actually gets committed and the other ones are to be considered "local" or "child" UoW feeding into the originator / parent. 

We're not doing a distributed transaction, but we're performing operations on items that are not necessarily in the same process. All we really want for the originating Unit of Work is to gather all the results into it before we do a commit. Once the commit is done or the UnitOfWork is completed we want to notify anything that was enlisted in the unit of work.
So in short, there is in reality only one UnitOfWork that matters, and we're only interested in gathering events that has occurred into this and then commit this.
